src/Hazelcast.Net/Clustering/Failover.cs

```diff
@@ -0,0 +1,164 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Hazelcast.Configuration;
+using Hazelcast.Core;
+using Microsoft.Extensions.Logging;
+
+namespace Hazelcast.Clustering
+{
+    /// <summary>
+    /// Failover class holds given alternative failover cluster options. 
+    /// It switches the cluster options by listening <see cref="OnClusterStateChanged"/> event handler. On each call of the event, 
+    /// tracks the number of trials and switches the current cluster <see cref="CurrentClusterOptions"/> to next one.
+    /// </summary>
+    internal class Failover
+    {
+        private readonly ILogger _logger;
+        private readonly ClusterState _state;
+        private readonly IEnumerable<ClusterOptions> _clusters;
+        private readonly IEnumerator<ClusterOptions> _clusterEnumerator;
+        private int _currentTryCount;
+        private readonly int _maxTryCount;
+        private bool _isEnabled;
+        private Action<ClusterOptions> _clusterOptionsChanged;
+
+        internal Failover(ClusterState state, IHazelcastOptions options, ILoggerFactory loggerFactory)
+        {
+            if (options == null || options.Failover == null) throw new ArgumentNullException(nameof(options));
+            if (loggerFactory == null) throw new ArgumentNullException(nameof(loggerFactory));
+            if (state == null) throw new ArgumentNullException(nameof(state));
+
+            _logger = loggerFactory.CreateLogger<Failover>();
+            _state = state;
+            _isEnabled = options.Failover.Enabled;
+
+            if (!options.Failover.Clusters.Any() && options.Failover.Enabled)
+            {
+                throw new ConfigurationException("Failover is enabled but there is no cluster(s) provided. Please, check your configurations.");
+            }
+
+            //main cluster config is also an alternative cluster in a circular list (->blue->green->)
+            var clusterList = new List<ClusterOptions>
+            {
+                new ClusterOptions()
+                {
+                    Authentication = options.Authentication,
+                    ClusterName = options.ClusterName,
+                    Heartbeat = options.Heartbeat,
+                    LoadBalancer = options.LoadBalancer,
+                    Networking = options.Networking,
+                    WaitForConnectionMilliseconds = options.WaitForConnectionMilliseconds
+                }
+            };
+            clusterList.AddRange(options.Failover.Clusters);
+            _clusters = clusterList;
+
+            _clusterEnumerator = _clusters.GetEnumerator();
+            ResetToFirstCluster();
+
+            _isEnabled = options.Failover.Enabled;
```

Comment:
duplicate, you've done it already above + if not enabled, why initialize it all?

Reply:
Good catch, fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/587/commits/66b28e3e6d08210f07fe686b7eee389cfdcc3665